Add RVV optimizations for SIMD-based calculations in ScalarQuantizer.cpp

vsvnakers opened the pull request:
We have added the RVV implementation code for RISCV, which is benchmarked against NEON, in scalarquantizer.cpp for performance testing. The decode_8_components of CPU and RVV were tested and scored on the RISCV board, and the implementation of RVV showed significant performance improvement.

One of the RVV tests regarding decode_8_components

### Test Principle:

This program compares the execution time of **Scalar decoding** and **RVV Vectorized decoding** to verify the performance improvements brought by RVV optimization.

1. **Scalar Decoding**: Each byte is processed individually and decoded sequentially.
2. **RVV Vectorized Decoding**: Uses RVV vector instructions to process 8 bytes at once, utilizing parallel processing to accelerate the decoding.

### Test Procedure:

1. **Initial Data**: Create a random array `code`.
2. **Scalar Decoding**: Decode each byte sequentially and measure the time.
3. **RVV Decoding**: Decode 8 bytes at once using RVV vector instructions and measure the time.
4. **Timing**: Record the time for Scalar decoding and RVV decoding.

### Output:

* **Dummy sum**: Used to ensure consistency in calculations.
* **RVV execution time**: Time taken for RVV decoding.
* **Scalar execution time**: Time taken for scalar decoding.

```
[root@EulixOS ~]# ./decode_8
Dummy sum = 9476415.0000
RVV time    = 1571.824 ms
Scalar time = 2450.067 ms
```

vsvnakers committed:
Add RVV optimizations for SIMD-based calculations in ScalarQuantizer.cpp

Signed-off-by: lyd1992 <liuyudong@iscas.ac.cn>

Signed-off-by: vsvnakers <vsnmakers@outlook.com>

vsvnakers commented:
I hope you’re all doing well! I wanted to kindly follow up on my PR (#4503), which has been open for about a week.  
This PR includes an optimization for `Scalarquantizer` using RVV, and I’d greatly appreciate it if one of potential maintainers(@mnorris11 @Enet4 @alexanderguzhva ) could take a look when you have a chance.  

If there are any issues, missing details, or areas that could be improved, please feel free to let me know — I’m more than happy to make adjustments.  
Thank you very much for your time and support in helping move this forward! 🙏

alexanderguzhva commented:
@vsvnakers my comments are the following
1. please remove unneeded debugging commented out code, such as `// printf(...);`
2. please remove unneeded meaningful commented out code, such as `// vfloat32m1x2_t res;`
3. please remove debugging code with active `printf(...);` lines
4. please keep existing code style, so there's no need to change `if (cond) { do_it(); }` into `if (cond) do_it();` for one-liners. The rules for formatting the code can be found in included `.clang-format` file(s)

overall, the change looks good

vsvnakers committed:
fix: cleanup debug code and follow project code style
Signed-off-by: lyd1992 <liuyudong@iscas.ac.cn>
Signed-off-by: vsvnakers <vsvnakers@outlook.com>

vsvnakers commented:
Hi @alexanderguzhva @mnorris11, thanks a lot for the detailed feedback 🙏
I’ve addressed the comments (removed the unnecessary debug/commented code and kept the existing code style as suggested).
When you have a moment, could you please take another look? I’d really appreciate it!

mnorris11 commented:
@vsvnakers , thanks for this contribution! After the large SIMD refactor, we can start reviewing it. ETA this month. @subhadeepkaran to check

mdouze commented:
Hi! AFAICS, the diff is still mainly formatting comments. Please use the clang formatter to make it readable. Thanks!

vsvnakers committed:
using clang-format -i ScalarQuantizer.cpp

Signed-off-by: lyd1992 <liuyudong@iscas.ac.cn>

Signed-off-by: vsvnakers <vsvnakers@outlook.com>

vsvnakers committed:
Normalize line endings to LF

Signed-off-by: lyd1992 <liuyudong@iscas.ac.cn>

Signed-off-by: vsvnakers <vsvnakers@outlook.com>

mnorris11 commented:
The refactor has complete. This can go into https://github.com/facebookresearch/faiss/blob/main/faiss/impl/scalar_quantizer/sq-rvv.cpp.
